pkg/server/delta/v3/server.go

```diff
@@ -44,6 +54,224 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			if more {
+				typ := resp.GetDeltaRequest().GetTypeUrl()
+
+				if isErr := resp == deltaErrorResponse; isErr {
+					return status.Errorf(codes.Unavailable, typ+" watch failed")
+				}
+
+				nonce, err := send(resp)
+				if err != nil {
+					return err
+				}
+
+				watch := watches.deltaResponses[typ]
+				watch.nonce = nonce
+				watches.deltaResponses[typ] = watch
+
+				state := watches.deltaStreamStates[typ]
+				if state.ResourceVersions == nil {
+					state.ResourceVersions = make(map[string]string)
+				}
+
+				state.ResourceVersions = resp.GetNextVersionMap()
+				watches.deltaStreamStates[typ] = state
+			}
+		case req, more := <-reqCh:
+			// input stream ended or errored out
+			if !more {
+				return nil
+			}
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			if s.callbacks != nil {
+				if err := s.callbacks.OnStreamDeltaRequest(streamID, req); err != nil {
+					return err
+				}
+			}
+
+			// we verify nonce only if nonce is not initialized
+			var nonce string
+			// the node information may only be set on the first incoming delta discovery request
+			if req.Node != nil {
+				node = req.Node
+				nonce = req.GetResponseNonce()
+			} else {
+				req.Node = node
+
+				// If we have no nonce, i.e. this is the first request on a delta stream, set one
+				nonce = strconv.FormatInt(streamNonce, 10)
+			}
+
+			// type URL is required for ADS but is implicit for xDS
+			if defaultTypeURL == resource.AnyType {
+				if req.TypeUrl == "" {
+					return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+				}
+			} else if req.TypeUrl == "" {
+				req.TypeUrl = defaultTypeURL
+			}
+
+			typeURL := req.GetTypeUrl()
+			var state stream.StreamState
+			// Initialize the state map if we haven't already.
+			// This means that we're handling the first request for this type on the stream.
+			if s, ok := watches.deltaStreamStates[typeURL]; !ok {
+				state.ResourceVersions = make(map[string]string)
+			} else {
+				state = s
+			}
+
+			// We are in the wildcard mode if the first request of a particular type has an empty subscription list
+			var found bool
+			if state.IsWildcard, found = isWildcard[typeURL]; !found {
+				state.IsWildcard = len(req.GetResourceNamesSubscribe()) == 0
+				isWildcard[typeURL] = state.IsWildcard
+			}
+
+			s.subscribe(req.GetResourceNamesSubscribe(), state.ResourceVersions)
+			// We assume this is the first request on the stream
+			if streamNonce == 0 {
+				for r, v := range req.InitialResourceVersions {
+					state.ResourceVersions[r] = v
+				}
+			} else if streamNonce > 0 && len(req.InitialResourceVersions) > 0 {
+				return status.Errorf(codes.InvalidArgument, "InitialResourceVersions can only be set on initial stream request")
+			}
+			s.unsubscribe(req.GetResourceNamesUnsubscribe(), state.ResourceVersions)
+
+			// cancel existing watch to (re-)request a newer version
+			watch, seen := watches.deltaResponses[typeURL]
```

Comment:
This PR uses at least three different names for this second return value for map lookup, maybe use `ok` consistently? I believe that's the most common name for this

Reply:
I like `ok` the best, I'll make the changes